### ISSUE TITLE
Trying to get SQL Formatting with external formatter to work 

### ISSUE THREAD

**Issue by Tix2garr** [dbeaver/dbeaver#8185]
MS SQL Server
DBeaver 7.0.0
Windows 10

I like using DBeaver.  My main "issue" with DBeaver is the formatting.  (Well also not being able to filter results if I have run more than one SQL statement i.e. "custom queries", but that's not the topic here.  <wry smile>)

My preferred SQL Server formatter is ApexSQL Refactor which is an add-in for Visual Studio/SSMS, though I prefer to work in DBeaver.  Many times I copy the script from DBeaver, paste it into SSMS, format the script, copy it then paste it back into DBeaver.  Obviously not an optimal method.

I recently found out that ApexSQL Refactor has a command line interface.  !!!!!  So, I set it up.  Well, I _tried_ to set it up...

A link to the [Refactor CLI page](https://knowledgebase.apexsql.com/apexsql-refactor-command-line-interface-cli-switches/) in case you're curious.

Here is what I used in a command window to test the formatting.  It read the FileForSQLRefactor.sql, formatted it, and wrote it back out to the same file successfully.

> "C:\Program Files\ApexSQL\ApexSQL Refactor\ApexSqlRefactor.exe" /is:"C:\Files\FileForSQLRefactor.sql" /os:"C:\Files" /frs /f

Parameters
/is: is the directory of the script(s) or can be the exact name of the file
/os: is the directory of the output (will default to the same filename)
/frs tells it we are formatting the script (alternative is /obfuscation)
/f tells it to force the overwrite of the file if it already exists

It seems that Refactor needs a file to work with so I checked the "Use temp file" checkbox and entered the following command:
> "C:\Program Files\ApexSQL\ApexSQL Refactor\ApexSqlRefactor.exe" /is:${file} /os:${file} /frs /f

I get a message back through ApexSQL Refactor (in the preferences window where the reformatted code would be) something like:

> Cannot create "C:\Users\mgrove\AppData\Local\temp\dbeaver-sql-format\<big long number\>.sql" because a file or directory with the same name already exists.

Oh, so it's trying to create a folder with the file's name, since I have the full filename as the output DIRECTORY.  Got it.  I tried:

> "C:\Program Files\ApexSQL\ApexSQL Refactor\ApexSqlRefactor.exe" /is:${file} /os:"C:\Users\mgrove\AppData\Local\Temp" /frs /f

and I get "The directory name is invalid."   ?!?!?!?!

Adding /test to the command shows the following:

> ApexSQL Refactor 2020.01.0406, (c) 2019 Quest Software Inc. ALL RIGHTS RESERVED.
> 
> formatting_script: True
> input_script: C:\Users\mgrove\AppData\Local\Temp\dbeaver-sql-format11014864956463797271sql
> output_script: C:\Users\mgrove\AppData\Local\Temp
> force: True
> test: True
> verbose: True
> Return codes: ...

I have tried various other things, but nothing seems to be playing nicely.  I looked in the DBeaver documentation, but it only mentions the standard formatting, nothing about the external formatter.  The closest I got was some forum entries here.

Anybody have any ideas?  <wry smile>

Thanks!

Michael

**Comment by aoeusnth** [comment]
I have the same problem but with Devart's SQL Complete for SQL Server. The CLI works but DBeaver appears to read the results from stdout which is sadly not an option with SQL Complete:

"C:\Program Files (x86)\Devart\dbForge SQL Complete\sqlcomplete.com" /formatsql /quiet /sourcefile:${file} /outputfile:${file}

**Comment by Tix2garr** [comment]
Well, since I never got any official response, I had to hack it to work.  There may have been a way to use DBeaver's format capability with a file to work with or something, but I got fed up with it and went a completely different direction.

**Additional programs I used:**

* Autohotkey - LOVE this program.  Is Windows-based.
* ApexSQL Refactor - Have used this for some time in SSMS, just recently found out it had command line, which is what made this possible!

You will need to install Refactor then go into SSMS to set it all up and getting the formatting version you want to work by default.

I have attached a copy of the autohotkey file and the command line parameters for ApexSQL Refactor.  You will need to rename the autohotkey file (Format SQL.txt) to (Format SQL.ahk).  The forum doesn't like the ahk extension...

Currently, it is set up 

* select the script you want to format 
* copy it into the clipboard
* hit Ctrl+Alt+Shift+F

A command window will show up, disappear and then your selected code should get replaced with the formatted version!  :-)

aoeusnth - Should work for you just fine as well, just replacing the program name and parameters on the RunWait line.

Let me know if anyone has any questions about any of this.  

Hope this helps!

Michael
[Format SQL.txt](https://github.com/dbeaver/dbeaver/files/4644448/Format.SQL.txt)
[RefactorHelp.txt](https://github.com/dbeaver/dbeaver/files/4644452/RefactorHelp.txt)

**Comment by aoeusnth** [comment]
> Well, since I never got any official response, I had to hack it to work. There may have been a way to use DBeaver's format capability with a file to work with or something, but I got fed up with it and went a completely different direction.
> 
> **Additional programs I used:**
> 
> * Autohotkey - LOVE this program.  Is Windows-based.
>
> ...
> 
> aoeusnth - Should work for you just fine as well, just replacing the program name and parameters on the RunWait line.

Thank you Michael for proposing the idea.

I implemented a version of this to work with SQLInForm's Java app (the executable version has a CLI compatible with DBeaver, but is sadly too buggy):

1. Copy the highlighted code to clipboard
2. Activate the SQLInForm Java app
3. Load a formatting profile and format the code
4. Paste the formatted code back into the original application

The AutoHotkey script is triggered by a key combination (programmed into my Kinesis Advantage2 board), so it can actually be used from any application, not just DBeaver — including Snowflake's Web GUI.

I've attached the AutoHotkey script below.

Thanks again,
KJ

[Format with SQLInForm (AHK Script).txt](https://github.com/dbeaver/dbeaver/files/5626731/Format.with.SQLInForm.AHK.Script.txt)
